src/gtest/test_checktransaction.cpp

```diff
@@ -1358,3 +1358,266 @@ TEST(ChecktransactionTests, CanopyAcceptsZeroVPubOld) {
+
+    // Shielded coinbase outputs must be recoverable with an all-zeroes ovk.
+    RawHDSeed rawSeed(32, 0);
+    GetRandBytes(rawSeed.data(), 32);
+    auto to = libzcash::OrchardSpendingKey::ForAccount(HDSeed(rawSeed), Params().BIP44CoinType(), 0)
+        .ToFullViewingKey()
+        .ToIncomingViewingKey()
+        .Address(0);
+    uint256 ovk;
+    builder.AddOutput(ovk, to, CAmount(1000), std::nullopt);
+
+    // orchard::Builder pads to two Actions, but does so using a "no OVK" policy for
+    // dummy outputs, which violates coinbase rules requiring all shielded outputs to
+    // be recoverable. We manually add a dummy output to sidestep this issue.
+    // TODO: If/when we have funding streams going to Orchard recipients, this dummy
+    // output can be removed.
+    builder.AddOutput(ovk, to, 0, std::nullopt);
+
+    auto bundle = builder
+        .Build().value()
+        .ProveAndSign({}, uint256()).value();
+
+    CMutableTransaction mtx;
+    mtx.fOverwintered = true;
+    mtx.nVersionGroupId = ZIP225_VERSION_GROUP_ID;
+    mtx.nVersion = ZIP225_TX_VERSION;
+    mtx.nConsensusBranchId = NetworkUpgradeInfo[Consensus::UPGRADE_NU5].nBranchId;
+
+    mtx.vin.resize(1);
+    mtx.vin[0].prevout.SetNull();
+    mtx.vin[0].scriptSig << 123;
+    mtx.orchardBundle = bundle;
+
+    CTransaction tx(mtx);
+    EXPECT_TRUE(tx.IsCoinBase());
+
+    // Write the transaction bytes out so we can modify them to test failure cases.
+    CDataStream ss(SER_DISK, PROTOCOL_VERSION);
+    ss << tx;
+
+    // Define some constants to use when calculating offsets to modify below.
+    const size_t HEADER_SIZE = 4 + 4 + 4 + 4 + 4;
+    const size_t TRANSPARENT_BUNDLE_SIZE = 1 + 32 + 4 + 1 + 2 + 4 + 1;
+    const size_t SAPLING_BUNDLE_SIZE = 1 + 1;
+    const size_t ORCHARD_BUNDLE_START = (HEADER_SIZE + TRANSPARENT_BUNDLE_SIZE + SAPLING_BUNDLE_SIZE);
+    const size_t ORCHARD_BUNDLE_VALUEBALANCE_OFFSET = (
+        ORCHARD_BUNDLE_START +
+        ZC_ZIP225_ORCHARD_NUM_ACTIONS_SIZE +
+        ZC_ZIP225_ORCHARD_ACTION_SIZE * 2 +
+        ZC_ZIP225_ORCHARD_FLAGS_SIZE);
+
+    // Verify the transaction is the expected size.
+    size_t txsize = ORCHARD_BUNDLE_START + ZC_ZIP225_ORCHARD_BASE_SIZE + ZC_ZIP225_ORCHARD_MARGINAL_SIZE * 2;
+    EXPECT_EQ(ss.size(), txsize);
+
+    // Coinbase transaction should fail non-contextual checks with valueBalanceSapling
+    // out of range.
+    {
+        std::vector<char> txBytes(ss.begin(), ss.end());
+        uint64_t valueBalanceBad = htole64(MAX_MONEY + 1);
+        std::copy((char*)&valueBalanceBad, (char*)&valueBalanceBad + 8, txBytes.data() + ORCHARD_BUNDLE_VALUEBALANCE_OFFSET);
+
+        CDataStream ssBad(txBytes, SER_DISK, PROTOCOL_VERSION);
+        CTransaction tx;
+        EXPECT_THROW((ssBad >> tx), std::ios_base::failure);
+
+        // We can't actually reach the CheckTransactionWithoutProofVerification
+        // consensus rule, because Rust is doing this validation at parse time.
```

Comment:
When the Rust code rejects the transaction, does it do so with the correct DoS call, i.e. `DoS(100, false, REJECT_INVALID, "bad-txns-valuebalance-toolarge", ...)`? It seems confusing and error-prone under maintenance to have two places in the code checking the same thing.

Reply:
We have no current way to detect this over the FFI.

Reply:
The core problem here is that we've moved from a by-constraint to a by-construction consensus system:
- The C++ code parses `valueBalanceSapling` as an integer with range `(-(1 << 64)..(1 << 64))` (i.e. `i64`), and then specifically constrains it to actually have range `[-MAX_MONEY..MAX_MONEY]`.
- The Rust code parses `valueBalanceSapling` and `valueBalanceOrchard` as integers with range `[-MAX_MONEY..MAX_MONEY]`.

Due to us now leveraging the Rust code for txid computation, we basically always reject at parse time anything we can (the reason we can still test the Sapling consensus rule is that we represent the data in C++, so we can just not do the txid computation; we can't do that for Orchard where we represent the data in Rust). The problem is the C++ code is not set up for this, as it assumes a lax type system and a by-constraint consensus system. This is also what gives rise to #5744: the C++ code assumes that a parse error emerging from the message-handling system is Just Bad, and does not have any differentiation between possible kinds of parsing errors.

Reply:
If we want to maintain existing consensus rule network error reporting, we'd need to augment the Rust errors to differentiate those cases, thread that information across the FFI, and then have specific handling for it on the C++ side. I wonder whether this is actually useful however, because all it would _functionally_ do is maintain some error codes in the network layer, but it would _semantically_ differentiate between "parser errors" and "constraint errors", when at the end of the day, they are all consensus rule failures.

Reply:
I don't know why we would need to have specific handling on a per-error basis on the C++ side: if we want to implement consensus rules in Rust, then we should report them with an error struct that includes the same information as would be passed to a `DoS` call, which then can be handled generically for all errors. Filed #5749.